Zend/zend_attributes.h

```diff
@@ -0,0 +1,82 @@
+#ifndef ZEND_ATTRIBUTES_H
+#define ZEND_ATTRIBUTES_H
+
+#define ZEND_ATTRIBUTE_TARGET_CLASS			1
+#define ZEND_ATTRIBUTE_TARGET_FUNCTION		2
+#define ZEND_ATTRIBUTE_TARGET_METHOD		4
+#define ZEND_ATTRIBUTE_TARGET_PROPERTY		8
+#define ZEND_ATTRIBUTE_TARGET_CLASS_CONST	16
+#define ZEND_ATTRIBUTE_TARGET_PARAMETER		32
+#define ZEND_ATTRIBUTE_TARGET_ALL			63
+
+#define ZEND_ATTRIBUTE_SIZE(argc) (sizeof(zend_attribute) + sizeof(zval) * (argc) - sizeof(zval))
+
+BEGIN_EXTERN_C()
+
+extern ZEND_API zend_class_entry *zend_ce_php_attribute;
+extern ZEND_API zend_class_entry *zend_ce_php_compiler_attribute;
+
+typedef struct _zend_attribute {
+	zend_string *name;
+	zend_string *lcname;
+	uint32_t offset;
```

Comment:
Maybe add a comment here what this offset means.

Reply:
Yes, and it might be more important to improve `zend_get_attribute` to avoid or clarify the offset parameter there.

Reply:
I think we should drop the `offset` argument from the function and introduce a new function `zend_get_parameter_attribute()`. This should be easier to use / understand. This way offset could start at 0 which is consistent with internal parameter offsets. The function will add 1 to the offset and do the lookup in the function attributes hash.